File: camerax_android/lib/src/camera2/interop/camera2_camera_info_impl.dart
import 'package:camerax_android/src/camera2.dart';
import 'package:camerax_android/src/api.dart';
import 'package:camerax_android/src/common.dart';
import 'package:camerax_android/src/core.dart';
import 'package:camerax_platform_interface/camerax_platform_interface.dart';

final class Camera2CameraInfoImpl implements Camera2CameraInfo {
  final Camera2CameraInfoProxyApi api;

  Camera2CameraInfoImpl.internal(this.api);

  @override
  Future<String> getCameraId() => api.getCameraId();

  @override
  Future<T?> getCameraCharacteristic<T>(CameraCharacteristics$Key<T> key) {
    final keyApi = key.api;
    if (keyApi is CameraCharacteristicsStringKeyProxyApi) {
      return api.getStringCameraCharacteristic(keyApi).then((e) => e as T?);
    }
    if (keyApi is CameraCharacteristicsIntKeyProxyApi) {
      return api.getIntCameraCharacteristic(keyApi).then((e) => e as T?);
    }
    if (keyApi is CameraCharacteristicsBooleanKeyProxyApi) {
      return api.getBooleanCameraCharacteristic(keyApi).then((e) => e as T?);
    }
    if (keyApi is CameraCharacteristicsIntArrayKeyProxyApi) {
      return api.getIntArrayCameraCharacteristic(keyApi).then((e) => e as T?);
    }
    if (keyApi is CameraCharacteristicsFloatArrayKeyProxyApi) {
      return api.getFloatArrayCameraCharacteristic(keyApi).then((e) => e as T?);
    }
    if (keyApi is CameraCharacteristicsIntRangeKeyProxyApi) {
      return api
          .getIntRangeCameraCharacteristic(keyApi)
          .then((e) => e?.impl as T?);
    }
    if (keyApi is CameraCharacteristicsLongRangeKeyProxyApi) {
      return api
          .getLongRangeCameraCharacteristic(keyApi)
          .then((e) => e?.impl as T?);
    }
    if (keyApi
        is CameraCharacteristicsCameraMetadataInfoSupportedHardwareLevelKeyProxyApi) {
      return api
          .getCameraMetadataInfoSupportedHardwareLevel(keyApi)
          .then((e) => e?.impl as T?);
    }
    if (keyApi
        is CameraCharacteristicsCameraMetadataControlModeArrayKeyProxyApi) {
      return api
          .getCameraMetadataControlModeArray(keyApi)
          .then((e) => e?.map((e) => e.impl).toList() as T?);
    }
    if (keyApi
        is CameraCharacteristicsCameraMetadataControlAeModeArrayKeyProxyApi) {
      return api
          .getCameraMetadataControlAeModeArray(keyApi)
          .then((e) => e?.map((e) => e.impl).toList() as T?);
    }
    if (keyApi
        is CameraCharacteristicsCameraMetadataControlAfModeArrayKeyProxyApi) {
      return api
          .getCameraMetadataControlAfModeArray(keyApi)
          .then((e) => e?.map((e) => e.impl).toList() as T?);
    }
    if (keyApi
        is CameraCharacteristicsCameraMetadataControlAwbModeArrayKeyProxyApi) {
      return api
          .getCameraMetadataControlAwbModeArray(keyApi)
          .then((e) => e?.map((e) => e.impl).toList() as T?);
    }
    throw TypeError();
  }
}

final class Camera2CameraInfoChannelImpl extends Camera2CameraInfoChannel {
  @override
  Camera2CameraInfo from(CameraInfo cameraInfo) {
    final api = Camera2CameraInfoProxyApi.from(cameraInfo: cameraInfo.api);
    return Camera2CameraInfoImpl.internal(api);
  }
}

File: camerax_android/android/src/main/kotlin/dev/zeekr/camerax_android/camera2/interop/Camera2CameraInfoImpl.kt
package dev.zeekr.camerax_android.camera2.interop

import androidx.annotation.OptIn
import androidx.camera.camera2.interop.Camera2CameraInfo
import androidx.camera.camera2.interop.ExperimentalCamera2Interop
import androidx.camera.core.CameraInfo
import dev.zeekr.camerax_android.CameraMetadataControlAeModeApi
import dev.zeekr.camerax_android.CameraMetadataControlAfModeApi
import dev.zeekr.camerax_android.CameraMetadataControlAwbModeApi
import dev.zeekr.camerax_android.CameraMetadataControlModeApi
import dev.zeekr.camerax_android.CameraMetadataInfoSupportedHardwareLevelApi
import dev.zeekr.camerax_android.CameraXApiPigeonProxyApiRegistrar
import dev.zeekr.camerax_android.PigeonApiCamera2CameraInfoProxyApi
import dev.zeekr.camerax_android.camera2.CameraCharacteristicsImpl
import dev.zeekr.camerax_android.camera2.cameraMetadataControlAeModeApi
import dev.zeekr.camerax_android.camera2.cameraMetadataControlAfModeApi
import dev.zeekr.camerax_android.camera2.cameraMetadataControlAwbModeApi
import dev.zeekr.camerax_android.camera2.cameraMetadataControlModeApi
import dev.zeekr.camerax_android.camera2.cameraMetadataInfoSupportedHardwareLevelApi
import dev.zeekr.camerax_android.common.IntRange
import dev.zeekr.camerax_android.common.LongRange

@OptIn(ExperimentalCamera2Interop::class)
class Camera2CameraInfoImpl(registrar: CameraXApiPigeonProxyApiRegistrar) :
    PigeonApiCamera2CameraInfoProxyApi(registrar) {
    override fun from(cameraInfo: CameraInfo): Camera2CameraInfo {
        return Camera2CameraInfo.from(cameraInfo)
    }

    override fun getCameraId(pigeon_instance: Camera2CameraInfo): String {
        return pigeon_instance.cameraId
    }

    override fun getStringCameraCharacteristic(
        pigeon_instance: Camera2CameraInfo, key: CameraCharacteristicsImpl.StringKey
    ): String? {
        return pigeon_instance.getCameraCharacteristic(key.instance)
    }

    override fun getIntCameraCharacteristic(
        pigeon_instance: Camera2CameraInfo, key: CameraCharacteristicsImpl.IntKey
    ): Long? {
        return pigeon_instance.getCameraCharacteristic(key.instance)?.toLong()
    }

    override fun getBooleanCameraCharacteristic(
        pigeon_instance: Camera2CameraInfo, key: CameraCharacteristicsImpl.BooleanKey
    ): Boolean? {
        return pigeon_instance.getCameraCharacteristic(key.instance)
    }

    override fun getIntArrayCameraCharacteristic(
        pigeon_instance: Camera2CameraInfo, key: CameraCharacteristicsImpl.IntArrayKey
    ): List<Long>? {
        return pigeon_instance.getCameraCharacteristic(key.instance)?.map { it.toLong() }
    }

    override fun getFloatArrayCameraCharacteristic(
        pigeon_instance: Camera2CameraInfo, key: CameraCharacteristicsImpl.FloatArrayKey
    ): List<Double>? {
        return pigeon_instance.getCameraCharacteristic(key.instance)?.map { it.toDouble() }
    }

    override fun getIntRangeCameraCharacteristic(
        pigeon_instance: Camera2CameraInfo, key: CameraCharacteristicsImpl.IntRangeKey
    ): IntRange? {
        val value = pigeon_instance.getCameraCharacteristic(key.instance)
        return if (value == null) null
        else IntRange(value)
    }

    override fun getLongRangeCameraCharacteristic(
        pigeon_instance: Camera2CameraInfo, key: CameraCharacteristicsImpl.LongRangeKey
    ): LongRange? {
        val value = pigeon_instance.getCameraCharacteristic(key.instance)
        return if (value == null) null
        else LongRange(value)
    }

    override fun getCameraMetadataControlModeArray(
        pigeon_instance: Camera2CameraInfo, key: CameraCharacteristicsImpl.CameraMetadataControlModeArrayKey
    ): List<CameraMetadataControlModeApi>? {
        return pigeon_instance.getCameraCharacteristic(key.instance)?.map { it.cameraMetadataControlModeApi }
    }

    override fun getCameraMetadataControlAeModeArray(
        pigeon_instance: Camera2CameraInfo, key: CameraCharacteristicsImpl.CameraMetadataControlAeModeArrayKey
    ): List<CameraMetadataControlAeModeApi>? {
        return pigeon_instance.getCameraCharacteristic(key.instance)?.map { it.cameraMetadataControlAeModeApi }
    }

    override fun getCameraMetadataControlAfModeArray(
        pigeon_instance: Camera2CameraInfo, key: CameraCharacteristicsImpl.CameraMetadataControlAfModeArrayKey
    ): List<CameraMetadataControlAfModeApi>? {
        return pigeon_instance.getCameraCharacteristic(key.instance)?.map { it.cameraMetadataControlAfModeApi }
    }

    override fun getCameraMetadataControlAwbModeArray(
        pigeon_instance: Camera2CameraInfo, key: CameraCharacteristicsImpl.CameraMetadataControlAwbModeArrayKey
    ): List<CameraMetadataControlAwbModeApi>? {
        return pigeon_instance.getCameraCharacteristic(key.instance)?.map { it.cameraMetadataControlAwbModeApi }
    }

    override fun getCameraMetadataInfoSupportedHardwareLevel(
        pigeon_instance: Camera2CameraInfo, key: CameraCharacteristicsImpl.CameraMetadataInfoSupportedHardwareLevelKey
    ): CameraMetadataInfoSupportedHardwareLevelApi? {
        return pigeon_instance.getCameraCharacteristic(key.instance)?.cameraMetadataInfoSupportedHardwareLevelApi
    }
}